### ISSUE TITLE
Setup Commands not running with Ray Cluster on GCP

### ISSUE THREAD

**Issue by arandhaw** [ray-project/ray#46451]
### What happened + What you expected to happen

I have been creating Ray Clusters on cloud VM's in Google Cloud. I've been having issues with the setup_commands in the ray cluster YAML file. These are supposed to run when new nodes are made.

The commands always run correctly on the head node. However, sometimes when new workers are created by the autoscaler, one or both of the worker nodes is not setup correctly. No errors appear in logs, but the worker is not set up correctly. It appears to work / stop working randomly.

The YAML file below is the configuration file I've been using. You'll need to change the <project-id> in 3 places for your specific cloud project. I've been creating the clusters using ray up <yaml file name> on the google cloud shell, then SSH'ing into the head node to run scripts. The error first started appearing when the autoscaler added more than one worker.

### Versions / Dependencies

Ray most recent version.
The cluster is created from the google cloud shell.

### Reproduction script
```
# An unique identifier for the head node and workers of this cluster.
cluster_name: gpu-cluster

# Cloud-provider specific configuration.
provider:
    type: gcp
    region: us-east1
    availability_zone: "us-east1-c"
    project_id: <the project ID>  # Globally unique project id
# The maximum number of workers nodes to launch in addition to the head
# node.
max_workers: 2

# The autoscaler will scale up the cluster faster with higher upscaling speed.
# E.g., if the task requires adding more nodes then autoscaler will gradually
# scale up the cluster in chunks of upscaling_speed*currently_running_nodes.
# This number should be > 0.
upscaling_speed: 2.0

# This executes all commands on all nodes in the docker container,
# and opens all the necessary ports to support the Ray cluster.
# Empty string means disabled.
# If a node is idle for this many minutes, it will be removed.
idle_timeout_minutes: 20
# How Ray will authenticate with newly launched nodes.
auth:
    ssh_user: ubuntu

# Tell the autoscaler the allowed node types and the resources they provide.
available_node_types:
    ray_head:
        # The resources provided by this node type.
        resources: {"CPU": 16}
        # Provider-specific config for the head node, e.g. instance type.
        node_config:
            machineType: n1-standard-16
            serviceAccounts:
              - email: "ray-autoscaler-sa-v1@<project name>.iam.gserviceaccount.com"
                scopes:
                 - "https://www.googleapis.com/auth/cloud-platform"
            disks:
              - boot: true
                autoDelete: true
                type: PERSISTENT
                initializeParams:
                  diskSizeGb: 100
                  # See https://cloud.google.com/compute/docs/images for more images
                  sourceImage: projects/ml-images/global/images/c0-deeplearning-common-cu121-v20231209-debian-11
            scheduling:
              - onHostMaintenance: TERMINATE

    ray_worker_gpu:
        # The minimum number of nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 0
        # The maximum number of workers nodes of this type to launch.
        # This takes precedence over min_workers.
        max_workers: 2
        # The resources provided by this node type.
        resources: {"CPU": 8, "GPU": 1}
        # Provider-specific config for the head node, e.g. instance type. By default
        # Ray will auto-configure unspecified fields such as subnets and ssh-keys.
        # For more documentation on available fields, see:
        # https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert
        node_config:
            machineType: n1-standard-8
            serviceAccounts:
              - email: "ray-autoscaler-sa-v1@<project-id>.iam.gserviceaccount.com"
                scopes:
                 - "https://www.googleapis.com/auth/cloud-platform"
            disks:
              - boot: true
                autoDelete: true
                type: PERSISTENT
                initializeParams:
                  diskSizeGb: 100
                  # See https://cloud.google.com/compute/docs/images for more images
                  sourceImage: projects/ml-images/global/images/c0-deeplearning-common-cu121-v20231209-debian-11
            # Make sure to set scheduling->onHostMaintenance to TERMINATE when GPUs are present
            # Run workers on preemtible instance by default.
            # Comment this out to use on-demand
            guestAccelerators:
              - acceleratorType: nvidia-tesla-t4
                acceleratorCount: 1
            metadata:
              items:
                - key: install-nvidia-driver
                  value: "True"
            scheduling:
              - preemptible: true
              - onHostMaintenance: TERMINATE
# Specify the node type of the head node (as configured above).
head_node_type: ray_head

# Files or directories to copy to the head and worker nodes. The format is a
# dictionary from REMOTE_PATH: LOCAL_PATH, e.g.
file_mounts: {}
#    "/path1/on/remote/machine": "/path1/on/local/machine",
#    "/path2/on/remote/machine": "/path2/on/local/machine",


# run before setup commands (also outside of any docker containers)
initialization_commands:
 - 'echo "Setup Commands Started" >> /home/ubuntu/logs.txt 2>&1'

# List of shell commands to run to set up nodes.
setup_commands:
  - 'echo "Setup Commands Started" >> /home/ubuntu/logs.txt 2>&1'
  - "pip3 install torch >> /home/ubuntu/logs.txt 2>&1"
  - "pip3 install torchvision >> /home/ubuntu/logs.txt 2>&1"
  - "pip3 install Pillow >> /home/ubuntu/logs.txt 2>&1"
  - "pip3 install requests >> /home/ubuntu/logs.txt 2>&1"
  - "pip3 install Flask >> /home/ubuntu/logs.txt 2>&1"

# Custom commands that will be run on the head node after common setup.
head_setup_commands:
  - 'echo "Head Commands Started" >> /home/ubuntu/logs.txt 2>&1'
  - "pip install google-api-python-client==1.7.8"

# Custom commands that will be run on worker nodes after common setup.
worker_setup_commands:
  - 'echo "Worker command Started" >> /home/ubuntu/logs.txt 2>&1'

# Command to start ray on the head node. You don't need to change this.
head_start_ray_commands:
    - ray stop
    - >-
      ulimit -n 65536;
      ray start
      --head
      --port=6379
      --object-manager-port=8076
      --autoscaling-config=~/ray_bootstrap_config.yaml

# Command to start ray on worker nodes. You don't need to change this.
worker_start_ray_commands:
    - ray stop
    - >-
      ulimit -n 65536;
      ray start
      --address=$RAY_HEAD_IP:6379
      --object-manager-port=8076
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by jjyao** [comment]
@arandhaw when you say worker nodes is not setup correctly, what's they symptoms?

**Comment by g-goessel** [comment]
I am experiencing the same issue.
When I start 1 worker node at a time, it works fine.
If the autoscaler tries to start multiple nodes it will fail to run the setup commands. The result is that ray is not installed.

> 2024-07-09 11:57:41,321 INFO updater.py:452 -- [5/7] Initializing command runner
> 2024-07-09 11:57:41,321 INFO updater.py:498 -- [6/7] No setup commands to run.
> 2024-07-09 11:57:41,322 INFO updater.py:503 -- [7/7] Starting the Ray runtime
> 2024-07-09 11:57:41,322 VINFO command_runner.py:371 -- Running `export RAY_OVERRIDE_RESOURCES='{"CPU":1}';export RAY_HEAD_IP=10.128.0.62; ray stop`
> 2024-07-09 11:57:41,322 VVINFO command_runner.py:373 -- Full command is `ssh -tt -i ~/ray_bootstrap_key.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_24bf68e341/c022e6b155/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@10.128.15.193 bash --login -c -i 'source ~/.bashrc; export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (export RAY_OVERRIDE_RESOURCES='"'"'{"CPU":1}'"'"';export RAY_HEAD_IP=10.128.0.62; ray stop)'`
> 
> ==> /tmp/ray/session_latest/logs/monitor.log <==
> 2024-07-09 11:57:41,488 INFO discovery.py:873 -- URL being requested: POST https://compute.googleapis.com/compute/v1/projects/animated-bit-413114/zones/us-central1-a/instances/ray-gcp-99498f66b357d8db-worker-5b1c305c-compute/setLabels?alt=json
> 2024-07-09 11:57:41,791 INFO node.py:348 -- wait_for_compute_zone_operation: Waiting for operation operation-1720526261534-61ccf3ca540ad-b590f7ab-89ab5f7d to finish...
> 2024-07-09 11:57:41,792 INFO discovery.py:873 -- URL being requested: GET https://compute.googleapis.com/compute/v1/projects/animated-bit-413114/zones/us-central1-a/operations/operation-1720526261534-61ccf3ca540ad-b590f7ab-89ab5f7d?alt=json
> 2024-07-09 11:57:46,930 INFO discovery.py:873 -- URL being requested: GET https://compute.googleapis.com/compute/v1/projects/animated-bit-413114/zones/us-central1-a/operations/operation-1720526261534-61ccf3ca540ad-b590f7ab-89ab5f7d?alt=json
> 2024-07-09 11:57:47,058 INFO node.py:367 -- wait_for_compute_zone_operation: Operation operation-1720526261534-61ccf3ca540ad-b590f7ab-89ab5f7d finished.
> 
> ==> /tmp/ray/session_latest/logs/monitor.out <==
> 2024-07-09 11:57:47,058 ERR updater.py:171 -- New status: update-failed
> 2024-07-09 11:57:47,064 ERR updater.py:173 -- !!!
> 2024-07-09 11:57:47,064 VERR updater.py:183 -- Exception details: {'message': 'SSH command failed.'}
> 2024-07-09 11:57:47,065 ERR updater.py:185 -- Full traceback: Traceback (most recent call last):
>   File "/home/ubuntu/.local/share/pipx/venvs/ray/lib/python3.12/site-packages/ray/autoscaler/_private/updater.py", line 166, in run
>     self.do_update()
>   File "/home/ubuntu/.local/share/pipx/venvs/ray/lib/python3.12/site-packages/ray/autoscaler/_private/updater.py", line 531, in do_update
>     self.cmd_runner.run(
>   File "/home/ubuntu/.local/share/pipx/venvs/ray/lib/python3.12/site-packages/ray/autoscaler/_private/command_runner.py", line 379, in run
>     return self._run_helper(
>            ^^^^^^^^^^^^^^^^^
>   File "/home/ubuntu/.local/share/pipx/venvs/ray/lib/python3.12/site-packages/ray/autoscaler/_private/command_runner.py", line 298, in _run_helper
>     raise click.ClickException(fail_msg) from None
> click.exceptions.ClickException: SSH command failed.


My `setup_commands` block is not empty, and it was successful in installing ray on the head node.

**Comment by g-goessel** [comment]
I think the setup commands are only run on the first worked that is being created.

I'm saying that because using `ray monitor` revealed that my setup command was being run on exactly on host, and all the others failed. Eventually, all the worker nodes have been configured.

**Comment by arandhaw** [comment]
@jjyao let me clarify exactly what seems to occur. 
The problem occurs when the autoscaler creates new worker nodes.
What is supposed to happen is that during startup, the setup commands are run on the worker nodes.
Instead, what sometimes happens is that on some of the nodes, none of the setup commands are run. 

Since I install dependencies in the setup commands (e.g., "pip install torch"), my python scripts fail since none of the required libraries have been installed. 

To be clear, the problem is not that the commands are failing and raising error messages. They are not being run at all.

**Comment by arandhaw** [comment]
I may have solved mt problem. It turns out the version of ray installed on the head and worker nodes was 2.8.1, whereas the version on the cluster launcher was 2.32.0. I just assumed that ray would install itself on the head/worker nodes, but I think it was using an older version part of the VM image. By adding "pip install -U ray[all]" to the setup commands, it seems to have fixed the problem.

It would be nice if the documentation was clearer (or if a meaningful error message was given by ray).

**Comment by Gabriellgpc** [comment]
> pip install -U ray[all]

I'll test that as well. thanks!

**Comment by astron8t-voyagerx** [comment]
I'm having same issue here.
The last clue I found is that this doesn't happen until version 2.6.3, and it happens from 2.7.0 onwards.

**Comment by Gabriellgpc** [comment]
What I'm doing to handle that for now, is use the "startup-script", a field from GCP 
https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert

In the **node_config**

`            metadata:
              items:
                - key: startup-script
                  value: |
                    #!/bin/bash
                    sudo mkdir -p /fiftyone/datasets
                    <more commands here>
`
It's working as expected this way.

**Comment by astron8t-voyagerx** [comment]
@Gabriellgpc Thank you for sharing! That's a good workaround, but I need the "file_mounts" option to sync files across the cluster and it seems hard to achieve that with startup scripts.
I've been struggling with this issue since version 2.7.0, but no fix has appeared yet.

**Comment by Gabriellgpc** [comment]
@astron8t-voyagerx I'm using the startup command to do something similar actually, because I want to share large folder between nodes, and the file mount do not handle that, I'm using the GCP buckets to store my large files and mounting it on all the nodes using the gsfuse command.
So my node_config looks like that,
```
            metadata:
              items:
                - key: startup-script
                  value: |
                    #!/bin/bash
                    sudo mkdir -p /fiftyone/datasets
                    sudo gcsfuse --implicit-dirs -o allow_other -file-mode=777 -dir-mode=777 --only-dir datasets fiftyone /fiftyone/datasets
```

**Comment by hartikainen** [comment]
I'm observing something very similar to this. My setup is the following:

`cluster-config.yml`:
```yml
[...]

cluster_synced_files: []
file_mounts_sync_continuously: false
rsync_exclude: []
rsync_filter:
  - ".gitignore"

initialization_commands: []

setup_commands:
  - cd ~/code && bazelisk run //:create_venv  # Use `rulesuv` to create `.venv` with pip packages.
  - echo 'source ~/code/.venv/bin/activate' >> ~/.bashrc
  - python -m ensurepip && python -m pip install -U google-api-python-client

head_setup_commands: []
worker_setup_commands: []

head_start_ray_commands:
  - ray stop
  - >-
    ray start
    --head
    --port=6379
    --object-manager-port=8076
    --autoscaling-config=~/ray_bootstrap_config.yaml
    --disable-usage-stats

# Command to start ray on worker nodes. You don't need to change this.
worker_start_ray_commands:
  - ray stop
  - >-
    ray start
    --address=$RAY_HEAD_IP:6379
    --object-manager-port=8076
    --disable-usage-stats
```

And here's a snippet of `/tmp/ray/session_latest/logs/monitor*`:
```
2024-11-13 15:14:49,324 VINFO command_runner.py:371 -- Running `uptime`                                                                         15:54:04 [351/1849]
2024-11-13 15:14:49,324 VVINFO command_runner.py:373 -- Full command is `ssh -tt -i ~/ray_bootstrap_key.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/
null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_24bf68e3
41/aaa3e6b268/%C -o ControlPersist=10s -o ConnectTimeout=10s ubuntu@10.150.0.74 bash --login -c -i 'source ~/.bashrc; export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignor
e && (uptime)'`
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

 15:14:50 up 7 min,  1 user,  load average: 0.00, 0.00, 0.00
2024-11-13 15:14:50,402 SUCC updater.py:295 -- Success.
2024-11-13 15:14:50,402 INFO log_timer.py:25 -- NodeUpdater: ray-cluster-test-worker-60b525fc-compute: Got remote shell  [LogTimer=6672ms]
2024-11-13 15:14:54,984 INFO updater.py:339 -- New status: waiting-for-ssh
2024-11-13 15:14:54,984 INFO updater.py:277 -- [1/7] Waiting for SSH to become available
2024-11-13 15:14:54,984 INFO updater.py:281 -- Running `uptime` as a test.
2024-11-13 15:14:54,984 INFO updater.py:389 -- Updating cluster configuration. [hash=09ddcf9108642756fa08d76c833ac1f21e014382]
2024-11-13 15:14:54,985 INFO command_runner.py:204 -- Fetched IP: 10.150.0.83
2024-11-13 15:14:54,990 INFO log_timer.py:25 -- NodeUpdater: ray-cluster-test-worker-61eab44a-compute: Got IP  [LogTimer=6ms]
2024-11-13 15:14:54,990 VINFO command_runner.py:371 -- Running `uptime`
2024-11-13 15:14:54,991 VVINFO command_runner.py:373 -- Full command is `ssh -tt -i ~/ray_bootstrap_key.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/
null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_24bf68e3
41/aaa3e6b268/%C -o ControlPersist=10s -o ConnectTimeout=10s ubuntu@10.150.0.83 bash --login -c -i 'source ~/.bashrc; export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignor
e && (uptime)'`
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

 15:14:56 up 7 min,  1 user,  load average: 0.16, 0.07, 0.03
2024-11-13 15:14:56,050 SUCC updater.py:295 -- Success.
2024-11-13 15:14:56,050 INFO log_timer.py:25 -- NodeUpdater: ray-cluster-test-worker-61eab44a-compute: Got remote shell  [LogTimer=1066ms]
2024-11-13 15:15:00,523 INFO updater.py:396 -- New status: syncing-files
2024-11-13 15:15:00,523 INFO updater.py:389 -- Updating cluster configuration. [hash=09ddcf9108642756fa08d76c833ac1f21e014382]2024-11-13 15:15:00,523   INFO update
r.py:254 -- [2/7] Processing file mounts

2024-11-13 15:15:00,528 INFO updater.py:271 -- [3/7] No worker file mounts to sync
2024-11-13 15:15:06,116 INFO updater.py:396 -- New status: syncing-files
2024-11-13 15:15:06,121 INFO updater.py:254 -- [2/7] Processing file mounts
2024-11-13 15:15:06,122 INFO updater.py:271 -- [3/7] No worker file mounts to sync
2024-11-13 15:15:11,733 INFO updater.py:407 -- New status: setting-up
2024-11-13 15:15:11,740 INFO updater.py:448 -- [4/7] No initialization commands to run.
2024-11-13 15:15:11,740 INFO updater.py:452 -- [5/7] Initializing command runner
2024-11-13 15:15:11,740 INFO updater.py:498 -- [6/7] No setup commands to run.
2024-11-13 15:15:11,740 INFO updater.py:503 -- [7/7] Starting the Ray runtime
```

Here's another snippet of the monitor log:
```
[... more logs ...]

Warning: Permanently added '10.150.0.48' (ED25519) to the list of known hosts.
Warning: Permanently added '10.150.0.65' (ED25519) to the list of known hosts.
Shared connection to 10.150.0.48 closed.
Shared connection to 10.150.0.65 closed.
Warning: Permanently added '10.150.0.65' (ED25519) to the list of known hosts.
Shared connection to 10.150.0.65 closed.
Exception in thread Exception in thread Thread-207 (spawn_updater):
Thread-204 (spawn_updater)Traceback (most recent call last):
:
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12/
threading.py", line 1075, in _bootstrap_inner
Traceback (most recent call last):
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12/
threading.py", line 1075, in _bootstrap_inner
    self.run()
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12/
threading.py", line 1012, in run
    self.run()    self._target(*self._args, **self._kwargs)

  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12/
threading.py", line 1012, in run
  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/autoscaler.py", line 1343, in spawn_updater
    updater = NodeUpdaterThread(
       self._target(*self._args, **self._kwargs) 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/autoscaler.py", line 1343, in spawn_updater
        updater = NodeUpdaterThread(
           ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^
^^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/updater.py", line 569, in __init__
^^    ^NodeUpdater.__init__(self, *args, **kwargs)^
^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/updater.py", line 103, in __init__
^
  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/updater.py", line 569, in __init__
    self.cmd_runner = provider.get_command_runner(
         NodeUpdater.__init__(self, *args, **kwargs) 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/updater.py", line 103, in __init__
          self.cmd_runner = provider.get_command_runner( 
             ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^^^
^^^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/gcp/node_provider.py", line 334, in get_command_runner
^^    ^instance = resource.get_instance(node_id)^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
    File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/gcp/node_provider.py", line 334, in get_command_runner
     instance = resource.get_instance(node_id) 
  ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/gcp/node.py", line 454, in get_instance
^^^    .execute()
 ^ ^ ^  ^^^^^^^^^^^^^^
^^
  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/ray/autoscaler/_private/gcp/node.py", line 454, in get_instance
  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
        .execute()return wrapped(*args, **kwargs)

          ^ ^ ^ ^ ^ ^ ^^^^^^
^^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
^^^    ^return wrapped(*args, **kwargs)^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^
^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/http.py", line 923, in execute
^^    ^resp, content = _retry_request(^
^ ^ ^ ^ ^ ^ ^ ^   ^ ^ ^ ^ ^ ^ ^  ^ 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/http.py", line 923, in execute
^^^    ^resp, content = _retry_request(^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/http.py", line 222, in _retry_request
          raise exception 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/http.py", line 191, in _retry_request
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/http.py", line 191, in _retry_request
 ^^^    ^resp, content = http.request(uri, method, *args, **kwargs)^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
     File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/http.py", line 222, in _retry_request
       raise exception 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/googleapiclient/http.py", line 191, in _retry_request
  ^    ^resp, content = http.request(uri, method, *args, **kwargs)^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/google_auth_httplib2.py", line 218, in request
^^    ^response, content = self.http.request(^
^  ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/google_auth_httplib2.py", line 218, in request
 ^^    ^response, content = self.http.request(^
^ ^^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/httplib2/__init__.py", line 1724, in request
         (response, content) = self._request( 
         ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/httplib2/__init__.py", line 1724, in request
  ^^    ^(response, content) = self._request(^
^  ^ ^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/httplib2/__init__.py", line 1444, in _request
         (response, content) = self._conn_request(conn, request_uri, method, body, headers) 
                ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/httplib2/__init__.py", line 1444, in _request
   ^^^^^    ^(response, content) = self._conn_request(conn, request_uri, method, body, headers)^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/httplib2/__init__.py", line 1396, in _conn_request
^^^    ^response = conn.getresponse()^
^ ^^   ^^   ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^^^
^  File "/home/ubuntu/code/.venv/lib/python3.12/site-packages/httplib2/__init__.py", line 1396, in _conn_request
^^^^^^
  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12/http/client.py", line 1428, in getresponse
    response = conn.getresponse()
        response.begin() 
   File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/http/client.py", line 331, in begin
            version, status, reason = self._read_status() 
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/http/client.py", line 1428, in getresponse
      response.begin() 
   File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12/http/client.py", line 331, in begin
         version, status, reason = self._read_status() 
  ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ 
      File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3
.12/http/client.py", line 292, in _read_status
        line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1") 
  ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^
^  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/http/client.py", line 292, in _read_status
^^^^    ^line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^
^  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/socket.py", line 720, in readinto
^^^^^^^^^^^    ^^^^^^return self._sock.recv_into(b)
^^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/socket.py", line 720, in readinto
  ^^^^^    ^return self._sock.recv_into(b)^
^ ^^^ ^^ ^ ^ ^ ^ ^ ^^^^ ^ 
   File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/ssl.py", line 1251, in recv_into
^^^^    ^return self.read(nbytes, buffer)^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^^^
^  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/ssl.py", line 1251, in recv_into
^^^    ^return self.read(nbytes, buffer)^
         ^ ^^ ^^^^^^^^^^^^^^^^^^
^  File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/ssl.py", line 1103, in read
^^^^^^    ^return self._sslobj.read(len, buffer)^
^ ^ ^ ^ ^ ^ ^ ^ 
   File "/home/ubuntu/.cache/bazel/_bazel_ubuntu/da272e0d09e1d329650b5ce1f38bb920/external/rules_python~~python~python_3_12_x86_64-unknown-linux-gnu/lib/python3.12
/ssl.py", line 1103, in read
  ^    ^return self._sslobj.read(len, buffer)^
^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^ssl^^^.^SSLError: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:2570)^
^^^^^^^^
ssl.SSLError: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:2570)
Warning: Permanently added '10.150.0.69' (ED25519) to the list of known hosts.
Shared connection to 10.150.0.69 closed.
Shared connection to 10.150.0.69 closed.
Shared connection to 10.150.0.69 closed.
Shared connection to 10.150.0.69 closed.
Shared connection to 10.150.0.69 closed.

[...more logs...]
```

Which makes me suspect that this is perhaps some sort of threading issue?

Note that some of my workers run exactly as expected. This seems to be happening more often the more nodes I have and about a quarter or half of my nodes setup fine when I run with 32 nodes and it feels like this is happening less frequently when running with fewer nodes.

**Comment by richardliaw** [comment]
@jjyao maybe we could do some investigation here? seems odd to have this reliability issue

**Comment by mangohehe** [comment]
same issue here, any updates?

**Comment by hartikainen** [comment]
Any luck with this? It seems like the issue can now be pretty reliably reproduced. Does anyone happen to have any workaround for this? I wonder if adding some delays/sleeps somewhere could help the issue.

**Comment by MortalHappiness** [comment]
@hartikainen Did you try Ray version 2.38.0? I followed the steps I mentioned above and experimented by adding `ray[default]==2.37.0` and `ray[default]==2.38.0` in the `setup_commands`. It seems that the issue doesn't occur with version 2.38.0. You can try updating to 2.38.0 first, and if it works without issues, I'll go ahead and close this issue.

**Comment by hartikainen** [comment]
No, unfortunately, neither `ray==2.38.0` nor `ray==2.40.0` fixes this issue. I spent some time today looking into this and, as I mentioned above, I believe this is an issue caused by incorrect handling of [google-api-python-client](https://github.com/googleapis/google-api-python-client)-library, which is thread-unsafe as per the [documentation](https://googleapis.github.io/google-api-python-client/docs/thread_safety.html). I opened a draft PR at #49440 with an attempt to fix this.
